*: improve handling of permanent errors

erikgrinaker opened the pull request:
***: improve handling of permanent errors**

RPC errors are normally retried. However, in some cases the errors are
permanent such that retries are futile, and this can cause operations to
appear to hang as they keep retrying -- e.g. when running operations on
a decommissioned node. There is already some detection of permanent
errors, but it is incomplete.

This patch attempts to improve coverage of permanent errors, in
particular in the context of decommissioned nodes, and adds test cases
for these scenarios.

Release note: None

**roachpb: propagate gRPC Status errors across Error**

`roachpb.Error` uses `errors.EncodeError()` and `errors.DecodeError()`
to preserve the original structured error. Unfortunately, these did not
handle gRPC `Status` errors, resulting in a generic unstructured error
after decoding. Since this is used while propagating errors through the
KV layer, it could prevent detection of the original error type.

Support for gRPC Status encoding was added in cockroachdb/errors 1.8.3,
this patch registers the error en/decoder such that these errors are
preserved across `roachpb.Error`. A later patch will extend existing
error handling code to make better use of these.

Release note: None

Resolves #62233, resolves #61470. Touches #56208.

Decommissioned nodes will keep running a bunch of async processes that
try (and fail) to communicate with the cluster. These have not been
addressed here, see #62693.

erikgrinaker committed:
roachpb: propagate gRPC Status errors across Error

`roachpb.Error` uses `errors.EncodeError()` and `errors.DecodeError()`
to preserve the original structured error. Unfortunately, these did not
handle gRPC `Status` errors, resulting in a generic unstructured error
after decoding. Since this is used while propagating errors through the
KV layer, it could prevent detection of the original error type.

Support for gRPC Status encoding was added in cockroachdb/errors 1.8.3,
this patch registers the error en/decoder such that these errors are
preserved across `roachpb.Error`. A later patch will extend existing
error handling code to make better use of these.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62608)
<!-- Reviewable:end -->

tbg reviewed (commented):
Thanks! One comment about unit testing the propagation but basically LGTM once that's done.

NB: when you say "might resolve XXX" it will get closed on merge.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-:-MWeKR7L-65-8pd9yUdr:b-etu7y8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/roachpb/errors.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-MWeKoID76VQH5esbnTc:-MWeKoID76VQH5esbnTd:bntm3j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ba5497cbb7453e64065c44ae37b6667ce5e6986/pkg/roachpb/errors.go#L25)):*
> ```Go
> 	"github.com/cockroachdb/errors"
> 	"github.com/cockroachdb/errors/errorspb"
> 	_ "github.com/cockroachdb/errors/extgrpc" // Register EncodeError support for gRPC Status.
> ```

nit: `// register EncodeError support for gRPC Status`

Can you add a unit test that proves that whatever this does "works"?
___
*[pkg/server/connectivity_test.go, line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-MWeKhKT7iOSCkymAqe0:-MWeKhKT7iOSCkymAqe1:b-brqcx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cf9ef0a2b8fe6d9913c81e2db961e4c7084b12/pkg/server/connectivity_test.go#L400)):*
> ```Go
> 	//
> 	// TODO(erikgrinaker): until cockroachdb/errors preserves grpcstatus.Error
> 	// across errors.EncodeError() we'll have to disable this, since the return
> ```

Mind stressing this for `min(~10k iterations, 30m)` or thereabouts?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62608)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
Reviewable is down for me, replying here.

>Mind stressing this for min(~10k iterations, 30m) or thereabouts?

I let it and the acceptance test run for a couple of hours more, and they both eventually failed/hung. There's still some (rare) code paths that obscure the original error. I'll mark this as draft for now and do a full pass of error handling for #61470.

erikgrinaker committed:
*: improve handling of permanent errors

RPC errors are normally retried. However, in some cases the errors are
permanent such that retries are futile, and this can cause operations to
appear to hang as they keep retrying -- e.g. when running operations on
a decommissioned node. There is already some detection of permanent
errors, but it is incomplete.

This patch attempts to improve coverage of permanent errors, in
particular in the context of decommissioned nodes, and adds test cases
for these scenarios.

Release note: None

erikgrinaker reviewed (commented):
Pretty sure this resolves the acceptance test. The hang happened while writing the event record, which races with the node liveness entry propagation. In cases where it loses the race, it would hang due to internal RPC retries; now, the txn will return an error that will get logged but not propagated to the decommission caller. Recording this entry is best-effort due to the lack of txn support for liveness entries. This is also the case for the status entry removal.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-:-MX221AX0Vj8Dx3Swt6t:b35y9lv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[pkg/roachpb/errors.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-MWeKoID76VQH5esbnTc:-MX21ARqCyc_1iSFaZlT:b-9f087b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ba5497cbb7453e64065c44ae37b6667ce5e6986/pkg/roachpb/errors.go#L25)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: `// register EncodeError support for gRPC Status`

Can you add a unit test that proves that whatever this does "works"?
</blockquote></details>

Yep, done.
___
*[pkg/server/connectivity_test.go, line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-MWeKhKT7iOSCkymAqe0:-MX21De21m4Os7Q_BwDE:bh3bftk) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cf9ef0a2b8fe6d9913c81e2db961e4c7084b12/pkg/server/connectivity_test.go#L400)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Mind stressing this for `min(~10k iterations, 30m)` or thereabouts?
</blockquote></details>

I've added a bunch of other tests as well, and stressed each of them for 30m, as well as the acceptance test for 1h. The only failures I've seen are I/O timeouts and handshake timeouts caused by excessive load, stressing at lower parallelism resolves it (8 is generally fine, 16 isn't).


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 6 of 6 files at r2, 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62608)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/roachpb/errors_test.go, line 214 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-MX6iqdEBlz1uyjNa0N6:-MX6iqdEBlz1uyjNa0N7:b1tif5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/172a846d795abe229e545eb909253a84a54bc001/pkg/roachpb/errors_test.go#L214)):*
> ```Go
> 
> 	decoded, ok := status.FromError(goErr)
> 	require.True(t, ok, "expected gRPC status error, got %T: %v", goErr, goErr)
> ```

nit: there's `require.IsType` or something like that
Not worth another CI cycle, just keep it in mind for next time.
___
*[pkg/server/admin_test.go, line 2098 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-MX6jOsXExiDh3fgAgXy:-MX6jOsXExiDh3fgAgXz:b-yjbea2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c970f542d520b3b0d0db5c7390ff081377607c5d/pkg/server/admin_test.go#L2098)):*
> ```Go
> 		ReplicationMode: base.ReplicationManual, // saves time
> 		ServerArgs: base.TestServerArgs{
> 			Insecure: true, // allows admin client without seting up certs
> ```

nit: setting


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62608#-:-MX6k4mS82ki8G9ElVK2:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
bors r=tbg
